### ISSUE TITLE
Grouping by "year" results in numeric years instead of timestamps

### ISSUE THREAD

**Issue by tlrobinson** [metabase/metabase#1992]
All other date grouping granularities return a timestamp (e.x. `2016-02-23T00:00:27.126Z`) except for "year" which returns a number (e.x. `2016`). For consistency we should always return a timestamp.

This is currently manifesting itself in the `multiseries_charts_v1` branch as all "year" grouped times being parsed as date within a couple seconds of the unix epoch...

**Comment by camsaul** [comment]
This is by design, [see the MBQL reference](https://github.com/metabase/metabase/wiki/Query-Language-'98#relative-datetime-unit)

*  A year is considered by most databases to be an "extraction" operation just like getting the month of the year or day of the month. So this behavior is a more natural mapping to SQL and other query languages
*  [`YYYY` is a valid ISO-8601 date string](https://en.wikipedia.org/wiki/ISO_8601#Years)

**Comment by tlrobinson** [comment]
* I'm not sure the most "natural" mapping to databases is what we want. I would prefer consistency with the other operations.
* I think that's just saying the year part must be 4 digits. Unless moment.js is wrong: https://tonicdev.com/tlrobinson/iso8601-years But regardless, if you were intending to return a ISO-8601 date string then it should probably be a string, not a number.

**Comment by agilliland** [comment]
Yeah, year is definitely a special case in the sense that it's already the most coarse part of a date/time, so I can see how it's more debatable about how best to return it.

Regarding the output of our queries it feels like it would be more consistent if it returned a full iso8601 date/time string truncated to just the year (e.g. "2001-01-01T07:00:00.000Z") because all of our other date/time groupings (hour, day, week, month, quarter) return using this same exact format.

@salsakran any thoughts here?

**Comment by tlrobinson** [comment]
Also note that using the ISO 8601 will allow us to return the timezone the query was run using (once we return the correct timezone. See #1635)

**Comment by camsaul** [comment]
Why not just look at the metadata that comes back with the fields instead of all this hackery attempting to parse things are strings and what not? Trying to "guess" what the field is instead of looking at the metadata has caused lots of issues in the past.

I'm wary of having to spend a day or two rewriting the year handling for 11 drivers and updating all the associated unit tests when we should just look at the metadata instead.

**Comment by tlrobinson** [comment]
Ironically this broke because I *started* looking at the metadata. Previously year groupings were being parsed as plain numbers, but now that I'm checking that the field is `DateTimeField` (etc) it thinks it should be getting a timestamp but is actually getting a number.

I could add a special case like

```
if (col[0].unit === "year") {
  return moment(row[0], "YYYY");
} else {
  return moment(row[0], moment.ISO_8601);
}
```

That feels like the wrong solution, but if it's going to take a day or two to change it then I can do that.

**Comment by agilliland** [comment]
I'll also mention that not all clients will have the benefit of the metadata.  CSV downloads for instance only include the values themselves along with the column names.

**Comment by camsaul** [comment]
Then in that case I would assume most people would prefer looking at `2015` and `2016` instead of `2015-01-01T00:00:00Z` and `2016-01-01T00:00:00Z` which is why I chose the former in the first place.

As far as timestamps go, nothing will ever come back as a unix timestamp. Remember we have other date groupings like `month-of-year` and `day-of-month` that come back as integers too. Parsing those as UNIX timestamps would be just as bad. Realistically, we'll need some knowledge of how to handle different units on the frontend -- treat `year` as a year, etc.

**Comment by salsakran** [comment]
For CSV downloads, I think the iso8601 formatting is unreadable. 

I'm cool with it being the way to represent a point in time internally, but exposing it to end users is not a good thing. It's very hard to parse and kind of hideous

**Comment by salsakran** [comment]
Playing with the csv formats, I think we should format time extracts there the same way we format them in a tableview.

**Comment by tlrobinson** [comment]
I question whether `month-of-year` etc should even be returning a column with `base_type` of `DateTimeField`. They're effectively a calculated columns with the same name as the original, but the type is actually `IntegerField`.

**Comment by tlrobinson** [comment]
I'm going to whitelist the `unit`s that are known to be timestamps for now, but I think this issue should be remain open https://github.com/metabase/metabase/commit/d3278aca531bb7d504bf858f3d5849adfbecf47a

**Comment by camsaul** [comment]
The `base_type` of the Field has always been about the underlying representation of the Field -- so we know which options are appropriate to display in the QB and how to handle it in the QP. So I think it's a really bad idea to start returning Fields with incorrect metadata. Down the road at some point we'll inevitably end up with an `IntegerField` widget getting displayed for a `DateTimeField` because of the way it was bucketed. And we will be all like :cry: 

There's never been any promise that a `DateTimeField` is going to come back from the QP as a string. Everytime one comes back we need to look at its unit and go from there:

1.  Is it a `DateTimeField`?
2.  If yes, look at its `unit` and handle appropriately.

Otherwise, you'll need to check everything for a `unit` and the special case logic will be spread all over the place instead of just applying to a single type of field.

**Comment by agilliland** [comment]
To bring this back in a little bit, I think this is a natural topic to debate a little because it does represent a true corner case where both of our existing patterns for dealing with groupings on a date/time field seem appropriate.

To recap, those patterns are:
1. when grouping a date/time by a time period (hour, day, week, month, quarter) we return an iso8601 datetime string truncated to the beginning of the period.  (e.g. "2011-04-01T07:00:00.000Z")
2. when grouping a date/time by a time extract (minute-of-hour, hour-of-day, day-of-week, day-of-month, day-of-year, week-of-year, month-of-year, quarter-of-year) we return just the relevant extracted value as an integer or a string value.  (e.g. 7 or "Monday")

The complication with year is that it very reasonably fits into both of the above stated patterns and could be thought of as a time period OR a time extract.  So there is a claim to be made for it to be treated using either of the above patterns.

So assuming that we aren't going to provide an option for both, the question is really just which one we feel is more fitting.

**Comment by tlrobinson** [comment]
In my mind, "year" clearly fits in the former, since it can be treated as a timeseries.

Also it's basically the difference between rounding and modulo operations.

**Comment by camsaul** [comment]
Yup. [A pretty similar summary of *truncation* vs. *extraction* is in the MBQL documentation](https://github.com/metabase/metabase/wiki/Query-Language-'98#datetime-units).

I thought about these same issues when I added datetime bucketing to MBQL, and here's why I chose to make it an "extraction" operation: 

*  It more closely matches the operation provided by most database. They've spent a lot of time thinking about what *their* query languages should provide, so they probably all chose similarly for a good reason. 
*  I thought looking at the year like `2015` vs `2015-01-01T00:00:00.000Z` was preferable when the frontend wasn't doing any special parsing (like CSV downloads)

Arguments can be made for either option. But we've already went down one rabbit hole, and I don't think it makes sense to consider investing days of effort to switch to the other option, especially when the jury is out on whether that other option is preferable.

**Comment by camsaul** [comment]
Also FWIW as far as MBQL is concerned [`YYYY` strings are valid datetime strings](https://github.com/metabase/metabase/wiki/Query-Language-'98#iso-8601-datetime-literals). The backend code for parsing ISO-8601 strings handles `YYYY` strings just fine. So it might be the case that `moment.js` doesn't like them but I don't think the limitations of a 3rd-party library should dictate the design of MBQL.

**Comment by tlrobinson** [comment]
If we're going to use `2016` for year, then why don't we use `2016-01` for month, `2016-W01` for week, `2016-01-01` for day? Those are all valid ISO-8601 dates as well.

**Comment by camsaul** [comment]
@tlrobinson what would be the advantage of doing things that way? Seems like it would be a lot of hassle for little apparent benefit

**Comment by camsaul** [comment]
Implemented by #10561
